abi.md

```diff
@@ -0,0 +1,177 @@
+---
+author:
+- The FIRRTL Specification Contributors
+title: Specification for the FIRRTL Language
+date: \today
+# Options passed to the document class
+classoption:
+- 12pt
+# Link options
+colorlinks: true
+linkcolor: blue
+filecolor: magenta
+urlcolor: cyan
+toccolor: blue
+# General pandoc configuration
+toc: true
+numbersections: true
+# Header Setup
+pagestyle:
+  fancy: true
+# Margins
+geometry: margin=1in
+# pandoc-crossref
+autoSectionLabels: true
+figPrefix:
+  - Figure
+  - Figures
+eqnPrefix:
+  - Equation
+  - Equations
+tblPrefix:
+  - Table
+  - Tables
+lstPrefix:
+  - Listing
+  - Listings
+secPrefix:
+  - Section
+  - Sections
+# This 'lastDelim' option does not work...
+lastDelim: ", and"
+---
+
+# Introduction
+
+FIRRTL defines a language/IR for describing synchronous hardware circuits.  This
+document specifies the mapping of FIRRTL constructs to verilog in a manner 
+similar to an application binary interface (ABI) which enables predictability of
+the output of key constructs necessary for the interoperability between circuits 
+described in FIRRTL and between other languages and FIRRTL output.
+
+This document describes multiple versions of the ABI, specifically calling 
+specific changes in later versions.  It is expected that a conforming FIRRTL
+compile can lower to all specified ABIs.  This mechanism exist to allow 
+improved representations when using tools which have better verilog support and
+allow migration of existing development flows to significant representational 
+changes.
+
+# FIRRTL System Verilog Interface
+
+To use a circuit describe in FIRRTL in a predictable way, the mapping of certain 
+behaviors and boundary constructs in FIRRTL to System Verilog must be defined.  
+Where possible this ABI does not impose constraints on implementation, 
+concerning itself primarily with the boundaries of the circuit.
+
+## On Modules
+
+### The Circuit and Top Level
+
+The top level module, as specified by the circuit name, shall be present as a
+System Verilog module of the same name.  This firrtl module is considered a 
+"public" module and subject to the lowering constraints for public modules.
+
+### External Modules
+
+An external module may be presumed to exist following the lowering constraints
```

Comment:
Is it worth mentioning that external modules have the optional `defname` override for the Verilog name?

Reply:
yes, quite.